Домашнее задание к лекции «Базовые понятия статистики»

# Обязательная часть
Будем осуществлять работу с непростым набором данных о состоянии здоровья лошадей, испытывающих кишечные колики.

Задание 1. 
Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [254]:
import pandas as pd
import numpy as np

In [255]:
df = pd.read_csv('horse_data.csv', header=None, na_values='?')
df

0   1        2     3      4     5    6    7    8    9   ...    18    19  \
0    2.0   1   530101  38.5   66.0  28.0  3.0  3.0  NaN  2.0  ...  45.0   8.4   
1    1.0   1   534817  39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...  50.0  85.0   
2    2.0   1   530334  38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...  33.0   6.7   
3    1.0   9  5290409  39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...  48.0   7.2   
4    2.0   1   530255  37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...  74.0   7.4   
..   ...  ..      ...   ...    ...   ...  ...  ...  ...  ...  ...   ...   ...   
295  1.0   1   533886   NaN  120.0  70.0  4.0  NaN  4.0  2.0  ...  55.0  65.0   
296  2.0   1   527702  37.2   72.0  24.0  3.0  2.0  4.0  2.0  ...  44.0   NaN   
297  1.0   1   529386  37.5   72.0  30.0  4.0  3.0  4.0  1.0  ...  60.0   6.8   
298  1.0   1   530612  36.5  100.0  24.0  3.0  3.0  3.0  1.0  ...  50.0   6.0   
299  1.0   1   534618  37.2   40.0  20.0  NaN  NaN  NaN  NaN  ...  36.0  62.0   

      20   21   22  23     24  25  26  27  
0    NaN  NaN  2.0   2  11300   0   0   2  
1    2.0  2.0  3.0   2   2208   0   0   2  
2    NaN  NaN  1.0   2      0   0   0   1  
3    3.0  5.3  2.0   1   2208   0   0   1  
4    NaN  NaN  2.0   2   4300   0   0   2  
..   ...  ...  ...  ..    ...  ..  ..  ..  
295  NaN  NaN  3.0   2   3205   0   0   2  
296  3.0  3.3  3.0   1   2208   0   0   1  
297  NaN  NaN  2.0   1   3205   0   0   2  
298  3.0  3.4  1.0   1   2208   0   0   1  
299  1.0  1.0  3.0   2   6112   0   0   2  

[300 rows x 28 columns]

In [256]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       299 non-null    float64
 1   1       300 non-null    int64  
 2   2       300 non-null    int64  
 3   3       240 non-null    float64
 4   4       276 non-null    float64
 5   5       242 non-null    float64
 6   6       244 non-null    float64
 7   7       231 non-null    float64
 8   8       253 non-null    float64
 9   9       268 non-null    float64
 10  10      245 non-null    float64
 11  11      256 non-null    float64
 12  12      244 non-null    float64
 13  13      196 non-null    float64
 14  14      194 non-null    float64
 15  15      53 non-null     float64
 16  16      198 non-null    float64
 17  17      182 non-null    float64
 18  18      271 non-null    float64
 19  19      267 non-null    float64
 20  20      135 non-null    float64
 21  21      102 non-null    float64
 22  22

In [257]:
df.columns = ['new_name_1','age', 'new_name_3','rectal temperature',
              'pulse','respiratory rate', 'temperature of extremities',
              'peripheral pulse',
             'new_name_9','new_name_10', 'new_name_11','new_name_12',
             'new_name_13','new_name_14', 'new_name_15','new_name_16',
             'new_name_17','new_name_18', 'new_name_19','total protein',
             'new_name_21','new_name_22', 'outcome','new_name_24',
             'new_name_25','new_name_26', 'new_name_27','new_name_28']
df

new_name_1  age  new_name_3  rectal temperature  pulse  respiratory rate  \
0           2.0    1      530101                38.5   66.0              28.0   
1           1.0    1      534817                39.2   88.0              20.0   
2           2.0    1      530334                38.3   40.0              24.0   
3           1.0    9     5290409                39.1  164.0              84.0   
4           2.0    1      530255                37.3  104.0              35.0   
..          ...  ...         ...                 ...    ...               ...   
295         1.0    1      533886                 NaN  120.0              70.0   
296         2.0    1      527702                37.2   72.0              24.0   
297         1.0    1      529386                37.5   72.0              30.0   
298         1.0    1      530612                36.5  100.0              24.0   
299         1.0    1      534618                37.2   40.0              20.0   

     temperature of extremities  peripheral pulse  new_name_9  new_name_10  \
0                           3.0               3.0         NaN          2.0   
1                           NaN               NaN         4.0          1.0   
2                           1.0               1.0         3.0          1.0   
3                           4.0               1.0         6.0          2.0   
4                           NaN               NaN         6.0          2.0   
..                          ...               ...         ...          ...   
295                         4.0               NaN         4.0          2.0   
296                         3.0               2.0         4.0          2.0   
297                         4.0               3.0         4.0          1.0   
298                         3.0               3.0         3.0          1.0   
299                         NaN               NaN         NaN          NaN   

     ...  new_name_19  total protein  new_name_21  new_name_22  outcome  \
0    ...         45.0            8.4          NaN          NaN      2.0   
1    ...         50.0           85.0          2.0          2.0      3.0   
2    ...         33.0            6.7          NaN          NaN      1.0   
3    ...         48.0            7.2          3.0          5.3      2.0   
4    ...         74.0            7.4          NaN          NaN      2.0   
..   ...          ...            ...          ...          ...      ...   
295  ...         55.0           65.0          NaN          NaN      3.0   
296  ...         44.0            NaN          3.0          3.3      3.0   
297  ...         60.0            6.8          NaN          NaN      2.0   
298  ...         50.0            6.0          3.0          3.4      1.0   
299  ...         36.0           62.0          1.0          1.0      3.0   

     new_name_24  new_name_25  new_name_26  new_name_27  new_name_28  
0              2        11300            0            0            2  
1              2         2208            0            0            2  
2              2            0            0            0            1  
3              1         2208            0            0            1  
4              2         4300            0            0            2  
..           ...          ...          ...          ...          ...  
295            2         3205            0            0            2  
296            1         2208            0            0            1  
297            1         3205            0            0            2  
298            1         2208            0            0            1  
299            2         6112            0            0            2  

[300 rows x 28 columns]

new_name_1  age  new_name_3  rectal temperature  pulse  respiratory rate  \
0           2.0    1      530101                38.5   66.0              28.0   
1           1.0    1      534817                39.2   88.0              20.0   
2           2.0    1      530334                38.3   40.0              24.0   
3           1.0    9     5290409                39.1  164.0              84.0   
4           2.0    1      530255                37.3  104.0              35.0   
..          ...  ...         ...                 ...    ...               ...   
295         1.0    1      533886                 NaN  120.0              70.0   
296         2.0    1      527702                37.2   72.0              24.0   
297         1.0    1      529386                37.5   72.0              30.0   
298         1.0    1      530612                36.5  100.0              24.0   
299         1.0    1      534618                37.2   40.0              20.0   

     temperature of extremities  peripheral pulse  new_name_9  new_name_10  \
0                           3.0               3.0         NaN          2.0   
1                           NaN               NaN         4.0          1.0   
2                           1.0               1.0         3.0          1.0   
3                           4.0               1.0         6.0          2.0   
4                           NaN               NaN         6.0          2.0   
..                          ...               ...         ...          ...   
295                         4.0               NaN         4.0          2.0   
296                         3.0               2.0         4.0          2.0   
297                         4.0               3.0         4.0          1.0   
298                         3.0               3.0         3.0          1.0   
299                         NaN               NaN         NaN          NaN   

     ...  total protein  new_name_21  new_name_22  outcome  new_name_24  \
0    ...            8.4          NaN          NaN      2.0            2   
1    ...           85.0          2.0          2.0      3.0            2   
2    ...            6.7          NaN          NaN      1.0            2   
3    ...            7.2          3.0          5.3      2.0            1   
4    ...            7.4          NaN          NaN      2.0            2   
..   ...            ...          ...          ...      ...          ...   
295  ...           65.0          NaN          NaN      3.0            2   
296  ...            NaN          3.0          3.3      3.0            1   
297  ...            6.8          NaN          NaN      2.0            1   
298  ...            6.0          3.0          3.4      1.0            1   
299  ...           62.0          1.0          1.0      3.0            2   

     new_name_25  new_name_26  new_name_27  new_name_28   2  
0          11300            0            0            2 NaN  
1           2208            0            0            2 NaN  
2              0            0            0            1 NaN  
3           2208            0            0            1 NaN  
4           4300            0            0            2 NaN  
..           ...          ...          ...          ...  ..  
295         3205            0            0            2 NaN  
296         2208            0            0            1 NaN  
297         3205            0            0            2 NaN  
298         2208            0            0            1 NaN  
299         6112            0            0            2 NaN  

[300 rows x 29 columns]

In [259]:
df_new = df[['age','rectal temperature',
              'pulse','respiratory rate', 'temperature of extremities',
              'peripheral pulse','total protein','outcome',
             ]]
df_new

age  rectal temperature  pulse  respiratory rate  \
0      1                38.5   66.0              28.0   
1      1                39.2   88.0              20.0   
2      1                38.3   40.0              24.0   
3      9                39.1  164.0              84.0   
4      1                37.3  104.0              35.0   
..   ...                 ...    ...               ...   
295    1                 NaN  120.0              70.0   
296    1                37.2   72.0              24.0   
297    1                37.5   72.0              30.0   
298    1                36.5  100.0              24.0   
299    1                37.2   40.0              20.0   

     temperature of extremities  peripheral pulse  total protein  outcome  
0                           3.0               3.0            8.4      2.0  
1                           NaN               NaN           85.0      3.0  
2                           1.0               1.0            6.7      1.0  
3                           4.0               1.0            7.2      2.0  
4                           NaN               NaN            7.4      2.0  
..                          ...               ...            ...      ...  
295                         4.0               NaN           65.0      3.0  
296                         3.0               2.0            NaN      3.0  
297                         4.0               3.0            6.8      2.0  
298                         3.0               3.0            6.0      1.0  
299                         NaN               NaN           62.0      3.0  

[300 rows x 8 columns]

In [260]:
df_new.describe()

age  rectal temperature       pulse  respiratory rate  \
count  300.000000          240.000000  276.000000        242.000000   
mean     1.640000           38.167917   71.913043         30.417355   
std      2.173972            0.732289   28.630557         17.642231   
min      1.000000           35.400000   30.000000          8.000000   
25%      1.000000           37.800000   48.000000         18.500000   
50%      1.000000           38.200000   64.000000         24.500000   
75%      1.000000           38.500000   88.000000         36.000000   
max      9.000000           40.800000  184.000000         96.000000   

       temperature of extremities  peripheral pulse  total protein     outcome  
count                  244.000000        231.000000     267.000000  299.000000  
mean                     2.348361          2.017316      24.456929    1.551839  
std                      1.045054          1.042428      27.475009    0.737187  
min                      1.000000          1.000000       3.300000    1.000000  
25%                      1.000000          1.000000       6.500000    1.000000  
50%                      3.000000          2.000000       7.500000    1.000000  
75%                      3.000000          3.000000      57.000000    2.000000  
max                      4.000000          4.000000      89.000000    3.000000

# Задание 2. 
Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

Рассмотрим показатель Pulse. при расчете около 10% удалилось, что показало отклонение среднее арифметическое изменилось на 1,65, это примерно на 2,5% от изначальных показателей. Думаю, что это отклонение при удалении выбросов допустимое при max = 184.0 и min = 30.0

In [261]:
#возьмем 
print(df_new.pulse.mean())
print(df_new[(df_new.pulse > 40) & (df_new.pulse < 140)].pulse.mean())

71.91304347826087
72.49180327868852


In [262]:
print(df_new.pulse.median())
print(df_new[(df_new.pulse > 40) & (df_new.pulse < 140)].pulse.median())

64.0
66.0


In [263]:
print(df_new.pulse.round().mode()[0])

48.0


In [264]:
q1 = df_new.pulse.quantile(0.25)
q3 = df_new.pulse.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) #нижняя граница не выбросов
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df_new[df_new.pulse.between(lower_bound, upper_bound, inclusive=True)].sort_values('age')
remove_outliers

age  rectal temperature  pulse  respiratory rate  \
0      1                38.5   66.0              28.0   
189    1                38.3   60.0              16.0   
190    1                37.1   40.0               8.0   
192    1                38.2   48.0              18.0   
193    1                 NaN   60.0              48.0   
..   ...                 ...    ...               ...   
103    9                38.0  140.0              68.0   
244    9                38.2  124.0              88.0   
23     9                38.3  130.0              60.0   
212    9                38.9  120.0              30.0   
109    9                38.4   84.0              40.0   

     temperature of extremities  peripheral pulse  total protein  outcome  
0                           3.0               3.0            8.4      2.0  
189                         3.0               1.0            6.0      1.0  
190                         NaN               1.0            6.7      1.0  
192                         1.0               1.0           74.0      1.0  
193                         3.0               3.0            7.6      2.0  
..                          ...               ...            ...      ...  
103                         1.0               1.0            5.3      1.0  
244                         1.0               3.0            8.0      1.0  
23                          NaN               3.0           70.0      1.0  
212                         1.0               3.0            6.3      1.0  
109                         3.0               3.0            6.6      2.0  

[271 rows x 8 columns]

In [265]:
print(remove_outliers.pulse.mean())
print(remove_outliers.pulse.median())

70.25830258302582
64.0


In [266]:
print(df.pulse.mean() - remove_outliers.pulse.mean()) 
#среднее арифимитическое изменилась

1.6547408952350509


In [267]:
print(df.pulse.median() - remove_outliers.pulse.median())

0.0


# Задание 3. 
Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

Выбраны были данные по pulse, но это не очень хороший пример все-таки, так как
mode = 48.0, а mean = 64.0 и median = 71.9. при этом средний возрат mean = 2.6, а mean = 1.0 и median = 1.0. Когда максимальный возрат 9.0, а пульс 184.0. Очень сильное отклонение и нужно делить лошадей по возрасту и проводить оценку внутри возрастной группы. 

In [268]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age                         300 non-null    int64  
 1   rectal temperature          240 non-null    float64
 2   pulse                       276 non-null    float64
 3   respiratory rate            242 non-null    float64
 4   temperature of extremities  244 non-null    float64
 5   peripheral pulse            231 non-null    float64
 6   total protein               267 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [269]:
for i in df_new.columns:
    pct_missing = df_new[i].isnull().mean()
    print(f'{i} - {pct_missing :.1%}')

age - 0.0%
rectal temperature - 20.0%
pulse - 8.0%
respiratory rate - 19.3%
temperature of extremities - 18.7%
peripheral pulse - 23.0%
total protein - 11.0%
outcome - 0.3%


In [270]:
df_del = df_new.dropna(axis = 1).info()
df_del
#удаление всех столцов не подходит

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   age     300 non-null    int64
dtypes: int64(1)
memory usage: 2.5 KB


In [271]:
print(df_new['pulse'].mean())
print(df_new['pulse'].median())
print(df_new['pulse'].mode()[0])

print(df_new['pulse'].std())
print(df_new['pulse'].var())

71.91304347826087
64.0
48.0
28.630556660735003
819.7087747035575


In [272]:
df_del1 = df_new.dropna().info()
df_del1
#очень много теряем информации

<class 'pandas.core.frame.DataFrame'>
Int64Index: 144 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age                         144 non-null    int64  
 1   rectal temperature          144 non-null    float64
 2   pulse                       144 non-null    float64
 3   respiratory rate            144 non-null    float64
 4   temperature of extremities  144 non-null    float64
 5   peripheral pulse            144 non-null    float64
 6   total protein               144 non-null    float64
 7   outcome                     144 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 10.1 KB


In [273]:
df_gr_1=df_new.groupby('pulse').mean().sort_values('age', ascending=False).reset_index()
df_gr_1.head(10)

pulse  age  rectal temperature  respiratory rate  \
0  184.0  9.0               38.80              84.0   
1  164.0  9.0               39.10              84.0   
2  160.0  9.0               38.60              20.0   
3  150.0  9.0               38.90              61.0   
4  146.0  9.0               39.20              96.0   
5  136.0  9.0               38.10              48.0   
6  128.0  9.0                 NaN              36.0   
7  124.0  9.0               38.50              62.0   
8  140.0  5.0               36.70              46.0   
9  130.0  5.0               38.25              38.0   

   temperature of extremities  peripheral pulse  total protein  outcome  
0                         1.0               NaN           3.30      2.0  
1                         4.0               1.0           7.20      2.0  
2                         3.0               NaN            NaN      2.0  
3                         1.0               3.0           7.35      1.5  
4                         NaN               NaN            NaN      2.0  
5                         3.0               3.0           4.90      2.0  
6                         3.0               3.0           7.80      2.0  
7                         2.0               2.0           7.80      1.5  
8                         2.0               2.0          37.15      2.0  
9                         4.0               3.0          76.00      2.0

In [274]:
df_gr_1 = df_gr_1[['pulse', 'age']]
df_gr_1.head(10)

pulse  age
0  184.0  9.0
1  164.0  9.0
2  160.0  9.0
3  150.0  9.0
4  146.0  9.0
5  136.0  9.0
6  128.0  9.0
7  124.0  9.0
8  140.0  5.0
9  130.0  5.0

In [275]:
print(df_gr_1['pulse'].mean())
print(df_gr_1['pulse'].median())
print(df_gr_1['pulse'].mode()[0])

print(df_gr_1['pulse'].std())
print(df_gr_1['pulse'].var())

88.67307692307692
83.0
30.0
38.31076291400095
1467.7145550527905


In [276]:
print(df_gr_1['age'].mean())
print(df_gr_1['age'].median())
print(df_gr_1['age'].mode()[0])

print(df_gr_1['age'].std())
print(df_gr_1['age'].var())

2.5983682983682987
1.0
1.0
2.917248444514621
8.510338487022976


In [277]:

print(df_new['pulse'].mean())
print(df_new['pulse'].median())
print(df_new['pulse'].mode()[0])

print(df_new['pulse'].std())
print(df_new['pulse'].var())

71.91304347826087
64.0
48.0
28.630556660735003
819.7087747035575


In [ ]:
#замена пропусков на среднюю
df_new['pulse'].fillna(df_new['pulse'].median(), inplace=True)
df_new['pulse'].fillna(df_new['pulse'].median()).isna().sum()

In [194]:
print(df_new['pulse'].mean())
print(df_new['pulse'].median())
print(df_new['pulse'].mode()[0])

print(df_new['pulse'].std())
print(df_new['pulse'].var())

71.28
64.0
64.0
27.54154538888526
758.536722408027


In [279]:
df_new['pulse'].fillna('no_info').isna().sum()

0

In [280]:
print(df_new['pulse'].mean())
print(df_new['pulse'].median())
print(df_new['pulse'].mode()[0])

print(df_new['pulse'].std())
print(df_new['pulse'].var())

71.91304347826087
64.0
48.0
28.630556660735003
819.7087747035575


In [281]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age                         300 non-null    int64  
 1   rectal temperature          240 non-null    float64
 2   pulse                       276 non-null    float64
 3   respiratory rate            242 non-null    float64
 4   temperature of extremities  244 non-null    float64
 5   peripheral pulse            231 non-null    float64
 6   total protein               267 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [282]:
for i in df_new.columns:
    pct_missing = df_new[i].isnull().mean()
    print(f'{i} - {pct_missing :.1%}')

age - 0.0%
rectal temperature - 20.0%
pulse - 8.0%
respiratory rate - 19.3%
temperature of extremities - 18.7%
peripheral pulse - 23.0%
total protein - 11.0%
outcome - 0.3%


In [283]:
df_new.fillna(method='ffill',inplace=True)

C:\Users\Hastya\AppData\Roaming\Python\Python38\site-packages\pandas\core\frame.py:4317: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [284]:
df_new.fillna(method='bfill',inplace=True)

In [285]:
for i in df_new.columns:
    pct_missing = df_new[i].isnull().mean()
    print(f'{i} - {pct_missing :.1%}')

age - 0.0%
rectal temperature - 0.0%
pulse - 0.0%
respiratory rate - 0.0%
temperature of extremities - 0.0%
peripheral pulse - 0.0%
total protein - 0.0%
outcome - 0.0%


In [286]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age                         300 non-null    int64  
 1   rectal temperature          300 non-null    float64
 2   pulse                       300 non-null    float64
 3   respiratory rate            300 non-null    float64
 4   temperature of extremities  300 non-null    float64
 5   peripheral pulse            300 non-null    float64
 6   total protein               300 non-null    float64
 7   outcome                     300 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [287]:
df_new.head(10)

age  rectal temperature  pulse  respiratory rate  \
0    1                38.5   66.0              28.0   
1    1                39.2   88.0              20.0   
2    1                38.3   40.0              24.0   
3    9                39.1  164.0              84.0   
4    1                37.3  104.0              35.0   
5    1                37.3  104.0              35.0   
6    1                37.9   48.0              16.0   
7    1                37.9   60.0              16.0   
8    1                37.9   80.0              36.0   
9    9                38.3   90.0              36.0   

   temperature of extremities  peripheral pulse  total protein  outcome  
0                         3.0               3.0            8.4      2.0  
1                         3.0               3.0           85.0      3.0  
2                         1.0               1.0            6.7      1.0  
3                         4.0               1.0            7.2      2.0  
4                         4.0               1.0            7.4      2.0  
5                         2.0               1.0            7.4      1.0  
6                         1.0               1.0            7.0      1.0  
7                         3.0               1.0            8.3      2.0  
8                         3.0               4.0            6.2      3.0  
9                         1.0               4.0            6.2      1.0